tests/csapi/thread_notifications_test.go

```diff
@@ -0,0 +1,196 @@
+package csapi_tests
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/tidwall/gjson"
+
+	"github.com/matrix-org/complement/internal/b"
+	"github.com/matrix-org/complement/internal/client"
+	"github.com/matrix-org/complement/runtime"
+)
+
+func syncHasUnreadNotifs(roomID string, check func(gjson.Result, gjson.Result) bool) client.SyncCheckOpt {
+	return func(clientUserID string, topLevelSyncJSON gjson.Result) error {
+		unreadNotifications := topLevelSyncJSON.Get("rooms.join." + client.GjsonEscape(roomID) + ".unread_notifications")
+		unreadThreadNotifications := topLevelSyncJSON.Get("rooms.join." + client.GjsonEscape(roomID) + ".unread_thread_notifications")
+		if !unreadNotifications.Exists() {
+			return fmt.Errorf("syncHasUnreadNotifs(%s): missing notifications", roomID)
+		}
+		if check(unreadNotifications, unreadThreadNotifications) {
+			return nil
+		}
+		return fmt.Errorf("syncHasUnreadNotifs(%s): check function did not pass: %v / %v", roomID, unreadNotifications.Raw, unreadThreadNotifications.Raw)
+	}
+}
+
+// Test behavior of threaded receipts and notifications.
+//
+// 1. Send a series of messages, some of which are in threads.
+// 2. Send combinations of threaded and unthreaded receipts.
+// 3. Ensure the notification counts are updated appropriately.
+//
+// This sends four messages as alice creating a DAG somewhat like:
+//
+// A<--B<--C
+// ^
+// +---D
```

Comment:
"DAG somewhat like" set off my pedantry alarm! I think this is a bit naughty since complement will generate $D$ using $\{C\}$ as its set of prev events.

Maybe let's use the word timeline instead of DAG?

```suggestion
// This sends four messages as alice creating a timeline somewhat like:
//
// A<--B<--C    [thread]
// ^
// +---D        [main timeline]
```